OpenRA.Mods.Common/Widgets/WidgetUtils.cs

```diff
@@ -214,46 +215,78 @@ public static string FormatTimeSeconds(int seconds, bool leadingMinuteZero)
 			return $"{minutes:D}:{seconds % 60:D2}";
 		}
 
-		public static string WrapText(string text, int width, SpriteFont font)
+		static StringBuilder SplitLine(ReadOnlySpan<char> line, int width, SpriteFont font, StringBuilder sb)
 		{
-			var textSize = font.Measure(text);
-			if (textSize.X > width)
+			// Scan forwards until we find the last word that fits
+			// This guarantees a small bound on the amount of string we need to search before a linebreak
+			var start = 0;
+			var posPrevSpace = 0;
+			for (var i = 0; i < line.Length; ++i)
 			{
-				var lines = text.Split('\n').ToList();
-
-				for (var i = 0; i < lines.Count; i++)
+				if (line[i] == ' ')
 				{
-					var line = lines[i];
-					if (font.Measure(line).X <= width)
-						continue;
-
-					// Scan forwards until we find the last word that fits
-					// This guarantees a small bound on the amount of string we need to search before a linebreak
-					var start = 0;
-					while (true)
+					var length = i - start;
+					if (length > 0)
 					{
-						var spaceIndex = line.IndexOf(' ', start);
-						if (spaceIndex == -1)
-							break;
+						var fragment = line.Slice(start, length);
+						if (font.Measure(fragment).X > width)
+						{
+							sb.Append(line.Slice(start, posPrevSpace - start)).Append('\n');
+							start = posPrevSpace + 1;
+						}
+
+						posPrevSpace = i;
+					}
+				}
+			}
 
-						var fragmentWidth = font.Measure(line.Substring(0, spaceIndex)).X;
-						if (fragmentWidth > width)
-							break;
+			return sb.Append(line.Slice(start, line.Length - start));
+		}
 
-						start = spaceIndex + 1;
-					}
+		public static string WrapText(string text, int width, SpriteFont font)
+		{
+			var textSize = font.Measure(text);
+			if (textSize.X <= width)
+				return text;
+
+			// Set capacity to twice the length, randomly chosen.
+			var sb = new StringBuilder(text.Length * 2);
+			var span = text.AsSpan();
+			var lineStart = 0;
+			var i = 0;
+			var length = 0;
 
-					if (start > 0)
+			for (; i < span.Length; ++i)
+			{
+				var ch = span[i];
+				if (ch == '\n')
+				{
+					length = i - lineStart;
+					if (length > 0)
 					{
-						lines[i] = line.Substring(0, start - 1);
-						lines.Insert(i + 1, line.Substring(start));
+						var line = span.Slice(lineStart, length);
+						if (font.Measure(line).X > width)
```

Comment:
Using `font.Measure` will duplicate a lot of calculations by looping over all previous characters in the slice each iteration. We could potentially optimize this a lot by moving this helper method into `SpriteFont` and keeping a running accumulation of `GlyphInfo.Advance`?

Reply:
Agree.

The idea also occured to me to have a UIString object that also contains the measure. Assuming font sizes won't chance ingame. Why measure each time - assuming that each string is rendered at least 60 times each second.

It would also be better if a string that does need wrapping is replaced with the wrapped result - to avoid needing to wrap each call. Although wrapping is not used that much in game. Only when chat window is open - and when there is something to wrap.

Reply:
WrapText and Font.Measure should generally already be cached inside a `CachedTransform` or similar, so this shouldn't be happening often.

Reply:
The  line you commented is only execute once per line in a string.

You couldn't store the advance in SpriteFont - since it depends on the string? You'd have to store it in the (UI)String?

So only in the SplitLines you could gain something by maintaining the width of the last fragment and only measure the width of the part you add to the fragment - the width between two spaces. But to me it feels the method does not get called often enough to make it that complex.

What does happen though is that after `WrapText` `Measure` is called once more. The UIString would solve this. Or WrapText could perhaps return a tuple (string, measureWidth).

Reply:
There may be value in introducing a new type that wraps the font, string, and derived metrics, but this will be a huge job to do properly so please not here!